crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -167,15 +168,244 @@ pub unsafe trait SystemParam: Sized {
     ) -> Self::Item<'world, 'state>;
 }
 
+pub unsafe trait OptionalSystemParam: Sized {
```

Comment:
Top level doc strings would be nice. I'd link to the derive macro docs here, along with a basic explanation.